hera_cal/lst_stack/metrics.py

```diff
@@ -0,0 +1,386 @@
+
+        for k, v in rdc.items():
+            if k == 'data':
+                k = 'mean'
+            dct = {antpair + (pol,): v[:, i, :, j] for i, antpair in enumerate(antpairs) for j, pol in enumerate(pols)}
+            outdict[k] = kls(dct)
+
+        return cls(**outdict)
+
+    @property
+    def bls(self):
+        return self.mean.bls()
+
+
+def get_nightly_predicted_variance(
+    bl: tuple[int, int, str],
+    stacks: list[UVData],
+    auto_stats: LSTBinStats,
+    dt: float,
+    df: float,
+) -> list[np.ndarray]:
+    """
+    Compute the predicted variance from the autos over nights/freqs for a particular bl.
+
+    Output is a list with the length of the number of lst bins, and each element is
+    an array of shape (Nnights, Nfreqs).
+    """
+
+    auto = auto_stats.mean[(bl[0], bl[0], bl[2])] * auto_stats.mean[(bl[1], bl[1], bl[2])]
+
+    per_day_expected_var = [None] * len(stacks)
+    dtdf = dt * df
+
+    for i, stack in enumerate(stacks):
+        gf = stack.get_flags(bl)
+        per_day_expected_var[i] = np.abs(auto[i] / dtdf / stack.get_nsamples(bl))
+        per_day_expected_var[i][gf] = np.inf
+
+    return per_day_expected_var
+
+
+def get_zscores(
+    auto_stats: LSTBinStats,
+    cross_stats: LSTBinStats,
+    dt: float,
+    df: float,
+    stacks: list[UVData],
+    central: str = 'mean',
+    std: str = 'autos'
+) -> list[UVData]:
+    """
+    Obtain Z-scores as a list of UVData objects.
+    """
+    zstacks = [stack.copy() for stack in stacks]  # UVData objects
+
+    if central not in ("mean", "median"):
+        raise ValueError("central must be 'mean' or 'median'")
+
+    for bl in cross_stats.bls:
+        if std == 'autos':
+            variance = get_nightly_predicted_variance(bl, stacks, auto_stats, dt, df)
+        elif std in {'std', 'mad'}:
+            variance = np.abs(np.where(cross_stats.flags[bl], np.inf, getattr(cross_stats, std)[bl]))**2
+        else:
+            raise ValueError("std must be 'autos', 'std' or 'mad'")
+
+        for i, (zstack, v) in enumerate(zip(zstacks, variance)):
+            data = zstack.get_data(bl)
+            flg = zstack.get_flags(bl)
+
+            centre = getattr(cross_stats, central)[bl][i]
+
+            z = np.abs(data - centre)**2 / (v / 2)  # TODO: check that we need /2 for std, mad
+
+            z[flg] = np.nan
+
+            zstack.set_data(z[:, :, None], bl)
+
+    return zstacks
+
+
+def get_selected_bls(
+    bls: list[tuple[int, int, str]],
+    days_binned: DataContainer | None = None,
+    selectors=None,
+    min_days: int = 7
+):
+    if selectors is None:
+        selectors = []
+    elif callable(selectors):
+        selectors = [selectors]
+
+    if min_days:
+        selectors.append(lambda bl: bl in days_binned and (np.median(days_binned[bl]) >= min_days))
+
+    def select(bl):
+        return all(sel(bl) for sel in selectors)
+
+    return [bl for bl in bls if select(bl)]
+
+
+def _get_data_subset(
```

Comment:
So, to summarize my thoughts, and our slack-based discussion: this function takes in a UVData object (actually, now I've wrapped this UVData object in an `LSTStack`, which feels just like a UVData object, but has some extra methods that make it easier to deal with the specific format of our data). Anyhow, UVData objects are nice under the hood in that the data itself is just an array, so we _can_ do operations on the whole array. In particular, that is what this function does (I've updated the name to `downselect_zscores`). I didn't want to use the built-in UVData `select()` because it also has to update all the metadata and do checks etc., whereas I just want to return a raw array at the end, so this should be faster. 

I don't want to use DataContainers for the lst-stack, because that requires a reformat of the natural shape of the data that gets read in, and so is slower. I _am_ at this stage using DataContainers for the _reduced_ data (lst-averaged) because it's much smaller and you end up wanting to look at it per-baseline etc.